### PR TITLE
Overview: Add SSL cert column to domain list

### PR DESCRIPTION
**Pull request by chad1008**
Related to 8398-gh-Automattic/dotcom-forge

## Proposed Changes

Adds an SSL column to the Domains table. While the target was specifically the Hosting Overview, this table is used in other locations as well. After testing each, I've included the SSL column across the board, but am open to feedback if it makes anything feel too cluttered.

Design doc: R8yjT7mjG0Uw8h2GPyIY6R-fi-4041_65153

### Screenshot:

image

> [!NOTE]
> I applied the suggestion from the design feedback post to also add the green/orange treatment to the text (instead of just the icon) because I'm pretty sure we'll end up wanting to go that way. If we don't, it'll be an easy change.

## Why are these changes being made?

SSL status is an essential site metric that is currently buried behind different screens/settings. Making this more visible solves a customer pain point and brings our product more in line with our competitors.

## Testing Instructions
- Visit the [Calypso Live link for this branch](https://calypso.live/?image=registry.a8c.com/calypso/app:build-116088)
- Open `/overview` for any of your sites by clicking on the Hosting Overview button
- Append `?flags=hosting-overview-refinements` to the end of the url
- Confirm that the SSL column is present and reporting accurately
- Click on the SSL title to confirm the sorting order can be reversed (note: if your list of domains has only one status and then a `null` value (represented as `-` for free WPCOM provided domains) the sort will probably only work the first time. As far as I can tell, the sorting utility doesn't resort those `null` values after the initial click.
- Use the back button on your browser to return to `/overview`.
- Click on the `Domains` button in the left-hand sidebar to view the global domain list
- confirm that SSL values are being reported, and different states are styled correctly
- Test the sorting again. This time, because you're an a11n, you probably have enough variety to see the full sort effect.
- Locate or search for a site in the list that you know has a custom domain name, then select that domain
- In the sidebar, select `Upgrades > Domains` to view the domains list for this site.
- Confirm that SSL data is listed for each of the domains on the site, once again with `-` displayed for free domains that aren't eligible for SSL
- Finally, register a new domain, for your test site. Preferably a `.blog` because `.com`s resolve SSL certificates faster than I resolve an ice cream cone on a hot August day (which, for clarity, is _really_ fast).
- Once the new domain is registered, reload the Hosting Overview and find your new domain in the domains table. Confirm that it's reporting the `newly_registered` domain's SSL status as `pending`.

### CODE REVIEW

**Review by nuriapenya — CHANGES_REQUESTED** [review]
>I applied the suggestion from the design feedback post to also add the green/orange treatment to the text (instead of just the icon) because I'm pretty sure we'll end up wanting to go that way. If we don't, it'll be an easy change.

Perfect 👌🏽 I think it makes more sense. 

Is there any way to have SSL section opened by default when linking users to domains/manage/?

screen 2024-08-05 at 18 51 31

**Comment by chad1008** [comment]
> Is there any way to have SSL section opened by default when linking users to domains/manage/?

Possibly! I didn't see a mechanism for that after a look today, but there's probably a way to do it that I'm just not seeing yet.  I'll investigate a bit more and ask around a little to see what's possible! It's a good idea, I wish I'd thought of it myself.

**Comment by fredrikekelund** [comment]
There's no existing mechanism for automatically opening the `Domain security` accordion on page load, but we could add one. Passing a URL query param seems like a good approach. 

For the record, this is what the domain settings page looks like currently when `ssl_status: disabled`. The `Domain security` section is inconspicuous, so calling attention to it makes sense.

container-jolly-ardinghelli calypso live_domains_manage_f26d dev_edit_f26d dev(Medium laptop) (1)

**Comment by fredrikekelund** [comment]
FYI, I noticed an issue with the `Contact us` link on the domain settings page when testing this PR. Logged this issue: https://github.com/Automattic/dotcom-forge/issues/8611

File: packages/domains-table/src/domains-table/domains-table-ssl-cell.tsx
Comment (fredrikekelund) [inline packages/domains-table/src/domains-table/domains-table-ssl-cell.tsx]:
```suggestion
	sslStatus: DomainData[ 'ssl_status' ];
```
Minor, but it would make things easier to reference the canonical type definition directly.

File: packages/domains-table/src/domains-table/style.scss
Comment (fredrikekelund) [inline packages/domains-table/src/domains-table/style.scss]: Do we know when a domain has `ssl_status: disabled`? I'm not sure, but if it indicates an error state, maybe red would be a more appropriate color.

**Review by fredrikekelund — COMMENTED** [review]
I know you haven't marked this ready for review just yet, @chad1008, but I left some small comments. Looking good overall, though!

Visually, I also think it'd be nice to reduce the width of the SSL column. It's currently quite wide compared to the columns right of it with plenty of whitespace.

File: packages/domains-table/src/domains-table/style.scss
Comment (chad1008) [reply packages/domains-table/src/domains-table/style.scss]: Excellent question. I've just done a bit of poking around (but pulled away before I fell all the way down the rabbit hole 😆) and didn't find an answer, but I _do_ know we don't allow site owners to disabled SSL themselves... so an error state feels most likely to me. Especially since the only site in my own account with SSL `disabled` is due to an expired domain.

@nuriapenya, any objection to red from a design perspective? I'd probably go with `studio-red-50`, because that's the color used for errors in the `status` column of this table. I worry a little that it might not have enough contrast with the `--studio-orange-50` we use for `pending`, but that also matches a similar state in the `status` column, so it's probably a broader conversation if it's an issue:

image

In this screenshot, we can see SSL `Pending` in `orange-50`, `Disabled` in `red-50`. We can also see `red-50` in the Status column on the right.

**Comment by fredrikekelund** [comment]
This screenshot comes from https://github.com/Automattic/dotcom-forge/issues/8611. It shows that we use an orange color for the padlock on the domain settings page when `ssl_status = disabled`. It seems reasonable to align the colors in the domains table with that page.

image

**Comment by chad1008** [comment]
Ah, good find @fredrikekelund. In that case, maybe we just use the same orange for both `disabled` and `pending` ...which alleviates my contrast concerns as well!

**Comment by chad1008** [comment]
Ready for review!

- Now using orange for `pending` and `disabled` (we can swap back to gray if design deems it best)
- SSL accordion will automatically expand when the link is clicked

**Comment by delputnam** [comment]
FYI, we have an update to the domain security card that is currently in progress. It will provide specific actionable data when a certificate is pending and will allow the user to try provisioning the certificate again once they've addressed the issues.

**Review by nightnei — APPROVED** [review]
Only one UI thing - lets align the title of the column with icons: 
Screenshot 2024-08-08 at 13 05 56


And in general LGTM and works as expected 👍
Screenshot 2024-08-08 at 12 14 11
Screenshot 2024-08-08 at 12 14 30
Screenshot 2024-08-08 at 12 18 21
Screenshot 2024-08-08 at 12 58 59

**Comment by nightnei** [comment]
> FYI, we have an update to the domain security card that is currently in progress. It will provide specific actionable data when a certificate is pending and will allow the user to try provisioning the certificate again once they've addressed the issues.

@delputnam have I understood correctly, that - if a certificate wasn't provisioned due to some issues, we will allow user to address issues and manually request provisioning again?

File: client/my-sites/domains/domain-management/settings/index.tsx
Comment (fredrikekelund) [inline client/my-sites/domains/domain-management/settings/index.tsx]:
```suggestion
				expanded={ queryParams.get( 'ssl-open' ) === 'true' }
```
On L334, there's an existing example of an accordion being expanded on page load. Let's follow the same approach here and drop the use of `useUrlQueryParam`

File: packages/domains-table/src/domains-table/style.scss
Comment (fredrikekelund) [inline packages/domains-table/src/domains-table/style.scss:161]: It looks like these styles aren't targeting the button properly. For disabled and pending domains, the button is blue. 

It's not a huge deal, but I would also advise us to follow the same approach to modifier classes for the icon and the button (either an `is-<status>` class, or a `domains-table-row__ssl-status-button__<status>` BEM modifier class).

I would personally opt for the `is-<status>` class generally because of the brevity, but either way is totally fine by me. Maybe there's a case to be made that the BEM modifier class is better for consistency in this particular file.

File: packages/domains-table/src/domains-table/domains-table-ssl-cell.tsx
Comment (fredrikekelund) [reply packages/domains-table/src/domains-table/domains-table-ssl-cell.tsx]: Looks like we still need to handle this change

File: packages/domains-table/src/domains-table/domains-table-ssl-cell.tsx
Comment (fredrikekelund) [inline packages/domains-table/src/domains-table/domains-table-ssl-cell.tsx]:
```suggestion
					href={ `${ domainManagementLink }?ssl-open=true` }
```
Can we pass the `href` directly and make it a true link?

**Review by fredrikekelund — CHANGES_REQUESTED** [review]
I left a few small comments that should be addressed before landing this. It looks like there's also a failing unit test we need to fix. I haven't looked closely at it, but I'm guessing it's because the test is querying the nth column of the table, or something like that.

File: packages/domains-table/src/domains-table/style.scss
Comment (fredrikekelund) [reply packages/domains-table/src/domains-table/style.scss:161]: I see you already caught this in fea46167dd870d9e6ad9176f258ff6c208c1c179, @chad1008 👍

**Comment by delputnam** [comment]
> if a certificate wasn't provisioned due to some issues, we will allow user to address issues and manually request provisioning again?

@nightnei Yes, that is correct.  In addition the the automatic retries, we'll be letting the user make a manual retry sooner if they believe they have fixed the underlying issue.

File: packages/domains-table/src/domains-table/style.scss
Comment (chad1008) [reply packages/domains-table/src/domains-table/style.scss:161]: Thanks for spotting this, I hadn't meant to leave two different naming conventions in place. Initially I transitioned from BEM to `is-<status>` because BEM specificity was struggling to overcome the Button component's base styles. I also liked the brevity of it as well.

That said, I think your point about consistency is a good one. I've updated the styles to use BEM exclusively, and gotten them specific enough that we don't have any unexpected hover behavior. It's a bit more verbose, but I think consistency within the file is the way to go.

File: packages/domains-table/src/domains-table/domains-table-ssl-cell.tsx
Comment (chad1008) [reply packages/domains-table/src/domains-table/domains-table-ssl-cell.tsx]: Technically yes, but I don't think it's ideal.

Our cell's parent `DomainsTableRow` has its own click handler, which is what opens `domains/manage` for the various cells. If we want our cell to behave differently and add a url parameter, we need a separate click handler that can stop propagation, at which point the new click handler also has to handle navigating us to the URL with the added param.

All of which means that while passing an `href` _will_ render the button as a true link, the `href` we've passed will effectively be ignored in favor of the click handler. We'd have to put the URL in both places (click handler and `href`), but only one of them would actually control navigation, which I worry is going to be unpleasant for future devs looking at this code.

...That is of course unless there's away around all of that I'm not thinking of yet.

Comment (fredrikekelund) [reply packages/domains-table/src/domains-table/domains-table-ssl-cell.tsx]: Good point 👍 That's what we do [here](https://github.com/Automattic/wp-calypso/blob/b81d5e143a5c5110841cba6d7beedaea6a160fb5/packages/domains-table/src/domains-table/domains-table-email-indicator.tsx#L65-L71) and [here](https://github.com/Automattic/wp-calypso/blob/89e0ede9a17cc29fd843ac277cf0cc12fcd72e47/packages/domains-table/src/domains-table/domains-table-row.tsx#L130-L136), though, so I think it's reasonable to use that same logic in this case.

Comment (fredrikekelund) [reply packages/domains-table/src/domains-table/domains-table-ssl-cell.tsx]: Making this into a regular anchor element instead of a `<Button>` component would also reduce the spacing between the icon and the button, which is currently a bit excessive (I compared it with the design doc).

Comment (chad1008) [reply packages/domains-table/src/domains-table/domains-table-ssl-cell.tsx]: Oh yes, that's much better. Passing the `href` to the `Button` would render an anchor but introduce the problems I mentioned. Just using an `anchor` with an `onClick` to stop propagation avoids that issue entirely, thanks!

File: packages/domains-table/src/domains-table/__tests__/domains-table-row.tsx
Comment (chad1008) [inline packages/domains-table/src/domains-table/__tests__/domains-table-row.tsx]: I'd prefer not to rely on a class name here, but the cell doesn't have much else to target (no role or unique content)

**Comment by chad1008** [comment]
All of @fredrikekelund's feedback should be addressed, and I've made a small update to the column widths, so the new SSL status is better spaced on the hosting overview page (previously it was pushed over too close to the renewal column)

File: packages/domains-table/src/domains-table/__tests__/domains-table-row.tsx
Comment (fredrikekelund) [reply packages/domains-table/src/domains-table/__tests__/domains-table-row.tsx]: AFAICT, this test passes even without this change, but my more importantly, I say let's update the `ssl_status` mock prop in `packages/domains-table/src/test-utils.tsx` and add one or two tests here for the new SSL status column (instead of changing this existing test).

As a general comment, let's also avoid querying the DOM directly in Jest tests except when strictly necessary. Adding a `data-testid` attribute is the preferred way when using Testing Library (see [docs](https://testing-library.com/docs/))

File: packages/domains-table/src/domains-table/domains-table-row.tsx
Comment (fredrikekelund) [inline packages/domains-table/src/domains-table/domains-table-row.tsx]:
```suggestion
				if ( column.name === 'ssl' ) {
```
Not strictly necessary, but since we already check the `hosting-overview-refinements` config flag in `packages/domains-table/src/domains-table-header/columns.ts`, I think we might as well drop it here. Especially since we throw an error on L239 if a column doesn't map to any component.

File: packages/domains-table/src/domains-table/domains-table-row.tsx
Comment (fredrikekelund) [inline packages/domains-table/src/domains-table/domains-table-row.tsx]:
```suggestion
						<DomainsTableSSLCell
							key={ domain.domain + column.name }
							domainManagementLink={ domainManagementLink }
							sslStatus={ sslStatus }
						/>
```
Let's move the `<td>` inside `DomainsTableSSLCell` and drop the `div`. That'll make styling easier.

File: packages/domains-table/src/domains-table/domains-table-ssl-cell.tsx
Comment (fredrikekelund) [inline packages/domains-table/src/domains-table/domains-table-ssl-cell.tsx]:
```suggestion
		<td className="domains-table-row__ssl-cell">
```
See my previous comment

File: packages/domains-table/src/domains-table/domains-table-ssl-cell.tsx
Comment (fredrikekelund) [inline packages/domains-table/src/domains-table/domains-table-ssl-cell.tsx:41]:
```suggestion
					icon={ lock }
					size={ 18 }
```
Will improve styling. 18 instead of 16 aligns with the renewal icon, and also seems to be what's used in the Figma doc.

File: packages/domains-table/src/domains-table/style.scss
Comment (fredrikekelund) [inline packages/domains-table/src/domains-table/style.scss]:
```suggestion
	.domains-table-row__ssl-icon {
```
Not needed.

File: packages/domains-table/src/domains-table/style.scss
Comment (fredrikekelund) [inline packages/domains-table/src/domains-table/style.scss:149]:
```suggestion
	.domains-table-row__ssl-cell {
		gap: 4px;
	}

	.domains-table-row__ssl-cell .domains-table-row__ssl-status-button {
```
Aligns with the way we style other table cells (particularly the renewal status cell)

**Review by fredrikekelund — CHANGES_REQUESTED** [review]
Currently, the spacing between the icon and the text doesn't match the spacing in the `Expires / renews on` cell. I left a few comments to address this. 

Let's also look into the test.

File: packages/domains-table/src/domains-table/__tests__/domains-table-row.tsx
Comment (fredrikekelund) [reply packages/domains-table/src/domains-table/__tests__/domains-table-row.tsx]: cc @nightnei

Comment (chad1008) [reply packages/domains-table/src/domains-table/__tests__/domains-table-row.tsx]: Excellent - I've gotten so used (in the past) to avoiding `data-testid`s that they didn't pop into my head until you suggested it. This test has been updated with a `data-testid` so it will continue to pass after the feature flag is removed. As we discussed earlier, we can look into alternate methods in a follow up if needed.

I haven't added new tests yet, and will need to take another look in the morning. There's a bit of a wrinkle, because even if we update the default ssl value as suggested, and/or pass a custom value, the `sslStatus` isn't parsed the same way during these tests as it would be in production. Because we're using test data and not an actual domain, [`useSiteDomainsQuery` comes back `undefined`](https://github.com/Automattic/wp-calypso/blob/01858f4725f9c3eb1050bd9df4b99fa999efe3e5/packages/domains-table/src/use-domain-row.ts#L30-L38). This, in turn, means `sslStatus` is `undefined`, making testing less than ideal. I'll investigate more in tomorrow.

File: packages/domains-table/src/domains-table/domains-table-ssl-cell.tsx
Comment (fredrikekelund) [inline packages/domains-table/src/domains-table/domains-table-ssl-cell.tsx:28]:
```suggestion
			return translate( 'Error' );
```
On the domain settings page, we use different wording for this, namely `Problem with SSL certificate`. 

https://github.com/Automattic/wp-calypso/blob/01858f4725f9c3eb1050bd9df4b99fa999efe3e5/client/my-sites/domains/domain-management/settings/helpers.ts#L13

I feel like there's some ambiguity with `Disabled`, where it could be mistaken for a user setting (as in "you've disabled the SSL certificate"). Maybe `Error` would convey meaning more accurately?

File: packages/domains-table/src/domains-table/domains-table-ssl-cell.tsx
Comment (fredrikekelund) [inline packages/domains-table/src/domains-table/domains-table-ssl-cell.tsx]:
```suggestion
					onClick={ ( e ) => e.stopPropagation() }
```
We don't need to manually type this. Typescript intuits the correct type.

**Review by fredrikekelund — APPROVED** [review]
LGTM 👍 I pushed a commit with two minor tweaks – one for removing a redundant TS type, and one for adjusting the spacing between the icon and the text (I know I said 4px originally, but because the renewal icon is wider within the SVG viewbox than the padlock icon, reducing the gap yields a better visual result).

@chad1008, let's also update the copy for the `disabled` state if you agree that `Error` is better wording than `Disabled`. Apart from that, I think we're good to land this.

Let's circle back to add tests after this PR is merged.

**Comment by fredrikekelund** [comment]
@delputnam, are you referring to the `ssl_status` prop..? Because looking at the code below, it would appear we display `Problem with SSL certificate` on the domain settings page when `ssl_status === disabled`.

https://github.com/Automattic/wp-calypso/blob/01858f4725f9c3eb1050bd9df4b99fa999efe3e5/client/my-sites/domains/domain-management/settings/helpers.ts#L13

**Comment by delputnam** [comment]
@fredrikekelund Yes, I know. That current message is incorrect.  I'm not sure how this got out of sync with the back-end, but we're fixing all of that in https://github.com/Automattic/wp-calypso/pull/93138.

**Comment by fredrikekelund** [comment]
I see. Thanks for pinpointing that. That does complicate things… It looks like we can't rely solely on the `ssl_status` prop, but we need to consider a few other heuristics. It would be ideal to reuse some of the logic from #93138 to generate the right message here.

@chad1008, I suggest we land this PR and follow up with another PR that corrects the text once #93138 is merged. This change won't be visible to users after merging since it's gated behind a feature flag.

**Comment by delputnam** [comment]
@fredrikekelund I think that the 'active', 'pending', or 'disabled' statuses may be appropriate for the overview page.  The user will need to go to the domain management page to see what the details of the issue actually is and take action. Essentially, if the status is 'pending', then we want them to go view the domain management page.

**Comment by chad1008** [comment]
Based on what @delputnam has shared, it sounds like the best course is to go back to the original strings (active, pending, disabled) and I'll create an issue on Letero's board to follow up once https://github.com/Automattic/wp-calypso/pull/93138 is deployed. Sound good @fredrikekelund?

**Comment by fredrikekelund** [comment]
[This code](https://github.com/Automattic/wp-calypso/pull/93138/files#diff-72abf15f3ceaeb592be417e9903b1606866efe30db89b8df5103685c0c66a83aR76-R85) is what gave me pause (i.e., that `pending` doesn't indicate failure if the domain is newly registered).

**Comment by fredrikekelund** [comment]
cc @gius80 for any additional insight

**Comment by delputnam** [comment]
> `pending` doesn't indicate failure if the domain is newly registered

@fredrikekelund It does still indicate that we "failed" to issue a certificate, but we know that it might be because the domain isn't resolving yet.  I think we have a couple different options to handle this case for the domains list:

- We can add a "newly_registered" option for the `ssl_status`.  We already have enough information to calculate this on the back-end so it wouldn't make this call much more expensive than it already is.
- We could use the `registration_date` property that is already in the domain object and calculate if the domain is less than 30 minutes old if the ssl status is `pending`

Basically, I don't think we should call the new endpoint to get the ssl status for the domains list because it is an expensive call.  We should just use data that we already have in one way or another.

**Comment by fredrikekelund** [comment]
That's great. Thanks, @delputnam 👍 

The 1st option sounds elegant and clear to me. Have you considered implementing that before landing #93138? If so, we could hold off with this PR. However, if it's a week or more away, I'd suggest landing this PR first (by implementing your 2nd suggestion) and then circling back to this once the back-end has been updated.

**Comment by delputnam** [comment]
@fredrikekelund Yes, we can go ahead and add that newly_registered status in now.  We'll let you know once it's deployed.

**Review by nightnei — APPROVED** [review]
LGTM 👍
[Splice](https://github.com/Automattic/wp-calypso/pull/93229/files#diff-fa97b8b9db60dcf3692b8b3a8cd026a600f663d54c2d400156e2c429b845cb22R158) with `3,0` looks not reliable, but we will rallback it after removing the feature flag, so I am ok to go with it 👍

**Comment by nightnei** [comment]
Screenshot 2024-08-15 at 11 44 50
Screenshot 2024-08-15 at 11 45 19
Screenshot 2024-08-15 at 12 05 03

**Comment by chad1008** [comment]
Changes to account for the new `newly_registered` SSL status have been added, and testing instructions updated accordingly.

File: packages/domains-table/src/domains-table/domains-table-ssl-cell.tsx
Comment (fredrikekelund) [inline packages/domains-table/src/domains-table/domains-table-ssl-cell.tsx:18]: I realize now that if the end goal was to continue displaying `Pending` when the `ssl_status` is `newly_registered`, then we might have been able to land this earlier 🤷‍♂️ Anyway, aligning with the behavior in #93138 is the main goal, so this makes sense.

**Review by fredrikekelund — APPROVED** [review]
LGTM 👍

**Comment by a8ci18n** [comment]
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/16492853

Some locales (Hebrew, Japanese) have been temporarily machine-translated due to translator availability. All other translations are usually ready within a few days. Untranslated and machine-translated strings will be sent for translation next Monday and are expected to be completed by the following Friday.

Thank you @chad1008 for including a screenshot in the description! This is really helpful for our translators.

**Pull request merged**
